### PR TITLE
IEnumXxx helpers v2

### PR DESCRIPTION
**Pull request by asklar**
This is another [simplified] attempt at the range-based iterator helpers for IEnumXYZ-style interfaces (v1 can be found in #234).
Merging @jonwis's, @oldnewthing's and my approach into this PR.
Some notable deltas:
- copy ctor/operator= need to be exposed
- make_range needs to return something that has begin() and end(), not a pair

### CODE REVIEW

File: include/wil/com.h [hunk 1]
```diff
@@ -3002,6 +3002,135 @@ namespace wil
...
+        using Result = T;
+    };
+
+
+    template<typename Interface>
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        using smart_result = std::conditional_t<
+            std::is_pointer_v<Result> &&
```
Comment (jonwis) [inline include/wil/com.h:3027]: Consider moving the conditionals into this line, then the output types onto the next.

File: include/wil/com.h [hunk 2]
```diff
@@ -3002,6 +3002,135 @@ namespace wil
...
+            else 
+            {
+                THROW_IF_FAILED_MSG(hr, "Failed to get next");
+            }
+        }
+    }
+};
+
+template<typename IEnumXXX>
+struct iterator_range
```
Comment (jonwis) [inline include/wil/com.h:3107]: Should this type be declared inside make_range so it's not just hanging around visible here?

Similarly with the iterator<T> - if it's local to make_range then you have to obtain one from make_range...

File: include/wil/com.h [hunk 3]
```diff
@@ -3002,6 +3002,135 @@ namespace wil
...
+        return m_begin;
+    }
+
+    WI_NODISCARD constexpr auto end() const noexcept
+    {
+        return iterator<IEnumXXX>(nullptr);
+    }
+};
+
+template<typename IEnumXxx, typename TStoredType = typename wil::details::com_enumerator_traits<IEnumXxx>::smart_result>
```
Comment (jonwis) [inline include/wil/com.h:3126]: Should TStoredType be removed, or used in the method body?

File: tests/ComTests.cpp [hunk 4]
```diff
@@ -2819,4 +2822,72 @@ TEST_CASE("StreamTests::Saver", "[com][IStream]")
...
+    }
+    {
+        wil::com_ptr<IShellFolder> desktop;
+        REQUIRE_SUCCEEDED(::SHGetDesktopFolder(&desktop));
+        wil::com_ptr<IEnumIDList> enumIDList;
+        REQUIRE_SUCCEEDED(desktop->EnumObjects(nullptr, SHCONTF_NONFOLDERS, &enumIDList));
+        REQUIRE(enumIDList);
+        
+        auto count = 0;
+        for (auto pidl : wil::make_range(enumIDList.get()))
```
Comment (jonwis) [inline tests/ComTests.cpp:2880]: Consider also a test on a mock "zero element" enumerator.

File: include/wil/com.h (hunk 2 above)
Comment (asklar) [reply include/wil/com.h:3107]: I'll move the `iterator_range` type inside `make_range`. I'd like to keep the iterator itself visible - esp. as it allows the types in the iterator to be tested, lmk if you disagree.

File: tests/ComTests.cpp [hunk 5]
```diff
@@ -2819,4 +2822,127 @@ TEST_CASE("StreamTests::Saver", "[com][IStream]")
...
+        {
+            REQUIRE(muffin == nullptr);
+            count++;
+            break;
+        }
+        REQUIRE(count == 0);
+        
+    }
+    {
+#if (NTDDI_VERSION >= NTDDI_VISTA)
```
Comment (dmachaj) [inline tests/ComTests.cpp:2903]: This is already in an ifdef cpp17 block.  Can you even compile pre-Vista headers in C++ 17 mode?

Comment (asklar) [reply tests/ComTests.cpp:2903]: not sure but i don't think we should tie together/assume lang version x SDK version. One check is for one thing, the other is for another, if the combination of C++17 w/pre-Vista is invalid, there's no harm (this check just makes it clear that this API was added in Vista)

**Review by jonwis — APPROVED** [review]

**Comment by asklar** [comment]
@jonwis I don't have merge permissions so if you're ok with it, can you press the `<button/>`?

File: include/wil/com.h [hunk 6]
```diff
@@ -3002,6 +3002,135 @@ namespace wil
...
+        using Result = T;
+    };
+
+
+    template<typename Interface>
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use IFoo*.
+        using smart_result = std::conditional_t<std::is_pointer_v<Result> && std::is_base_of_v<::IUnknown, std::remove_pointer_t<Result>>,
```
Comment (dunhor) [inline include/wil/com.h:3034]: `wistd::*`

File: include/wil/com.h [hunk 7]
```diff
@@ -3002,6 +3002,135 @@ namespace wil
...
+        using Interface = Itf;
+        using Result = T;
+    };
+
+
+    template<typename Interface>
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use IFoo*.
```
Comment (dunhor) [inline include/wil/com.h:3033]:
```suggestion
        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use the raw type.
```
A little confusing

File: include/wil/com.h [hunk 8]
```diff
@@ -3002,6 +3002,135 @@ namespace wil
...
+        using smart_result = std::conditional_t<std::is_pointer_v<Result> && std::is_base_of_v<::IUnknown, std::remove_pointer_t<Result>>,
+            wil::com_ptr<std::remove_pointer_t<Result>>, Result>;
+    };
+}
+
+template <typename IEnumType, typename TStoredType = typename details::com_enumerator_traits<IEnumType>::smart_result>
+struct iterator
```
Comment (dunhor) [inline include/wil/com.h:3040]:
```suggestion
struct com_enum_iterator
```
Given that this is directly in the `wil` namespace, the name should be more specific

File: include/wil/com.h [hunk 9]
```diff
@@ -3002,6 +3002,135 @@ namespace wil
...
+            {
+                THROW_IF_FAILED_MSG(hr, "Failed to get next");
+            }
+        }
+    }
+};
+
+
+template<typename IEnumXxx, typename TStoredType = typename wil::details::com_enumerator_traits<IEnumXxx>::smart_result>
+WI_NODISCARD auto make_range(IEnumXxx* enumPtr)
```
Comment (dunhor) [inline include/wil/com.h:3108]: This needs either a new name (so it doesn't conflict with other `make_range` functions, or it needs SFINAE

Comment (asklar) [reply include/wil/com.h:3108]: ah got it, that makes sense. the other make_range overloads are:
- make_range(T begin, T end)
- make_range(T begin, size_t count)

This overload is make_range(T), so it shouldn't conflict as long as there isn't another one-param overload. Right?

Comment (dunhor) [reply include/wil/com.h:3108]: Isn't there `make_range(IVector<T>*)` (& company)?

Even so, I don't think those would conflict since they are more "specific" and would "win" in overload resolution. However, we should still future-proof where we can. Enabling this overload only when the type has `T::Next` (likely of the desired signature) is certainly doable.

Comment (dunhor) [reply include/wil/com.h:3108]: Nvm, that's `get_range(IVector<T>`, etc. Second comment still stands (and now I'm kind of sad we have two naming conventions here)

Comment (asklar) [reply include/wil/com.h:3108]: I'm not sure how you're proposing that we do it while at the same time having a default TStoredType param... did you have a proposal in mind?

Comment (dunhor) [reply include/wil/com.h:3108]: Is it required to make `TStoredType` a template argument? If not, the function body can deduce this type and it need not be part of the signature. Otherwise, you can do this [the classic SFINAE way](https://godbolt.org/z/881EE5h7x)

Comment (asklar) [reply include/wil/com.h:3122]: I've moved TStoredType to not be a template argument

File: include/wil/com.h [hunk 10]
```diff
@@ -3002,6 +3002,150 @@ namespace wil
...
+    struct has_next
+    {
+      template <typename U = T>
+      static auto test(int) -> decltype(wistd::declval<U>()->Next(0, nullptr, nullptr), wistd::true_type{});
+
+      template <typename>
+			static auto test(...) -> wistd::false_type;
+
+			static constexpr bool value = decltype(test<T>(0))::value;
+    };
```
Comment (dunhor) [inline include/wil/com.h:3038]:
```suggestion
    template<typename T>
    struct has_next
    {
        template <typename U = T>
        static auto test(int) -> decltype(wistd::declval<U>()->Next(0, nullptr, nullptr), wistd::true_type{});

        template <typename>
        static auto test(...) -> wistd::false_type;

        static constexpr bool value = decltype(test<T>(0))::value;
    };
```
Something wonky happened with the formatting here

File: include/wil/com.h [hunk 11]
```diff
@@ -3002,6 +3002,150 @@ namespace wil
...
+
+    template<typename T>
+    constexpr bool has_next_v = has_next<T>::value;
+
+    template<typename Interface, wistd::enable_if_t<has_next_v<Interface*>, int> = 0>
```
Comment (dunhor) [inline include/wil/com.h:3043]: nit: I don't think the `enable_if` is doing anything here?

**Comment by dunhor** [comment]
A couple additional notes that are not urgent as they can be added later if desired:
1. The iterator type lacks an `iterator_category` typedef (and other typedefs), which may limit its usability in STL or other algorithms. This should probably be an input iterator I think?
2. The iterator type lacks a postfix `operator++(int)` which may also limit its usability in algorithms, and is a requirement for input iterators

File: include/wil/com.h (hunk 11 above)
Comment (asklar) [reply include/wil/com.h:3043]: i figured we'd want to sfinae this too? can remove if it's not useful.

File: include/wil/com.h (hunk 10 above)
Comment (asklar) [reply include/wil/com.h:3038]: oh weird. I think that was probably ghcp using tabs instead of spaces :(

File: include/wil/com.h (hunk 11 above)
Comment (dunhor) [reply include/wil/com.h:3043]: SFINAE on class templates _can_ be useful if you've got a number of definitions fighting each other (I've used the `void_t` trick on a number of occasions for this). In this case, it's not all that useful since the traits type shouldn't ever be used with a non-IEnum* type

File: include/wil/com.h [hunk 12]
```diff
@@ -3002,6 +3002,150 @@ namespace wil
...
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use the raw pointer type IFoo*.
+        using smart_result = wistd::conditional_t<std::is_pointer_v<Result> && wistd::is_base_of_v<::IUnknown, wistd::remove_pointer_t<Result>>,
```
Comment (dunhor) [inline include/wil/com.h:3048]:
```suggestion
        using smart_result = wistd::conditional_t<wistd::is_pointer_v<Result> && wistd::is_base_of_v<::IUnknown, wistd::remove_pointer_t<Result>>,
```
Missed this one on prior review

**Review by dunhor — APPROVED** [review]

**Comment by asklar** [comment]
@dunhor going to write a couple more tests, will ping back in a sec :)

File: include/wil/com.h [hunk 13]
```diff
@@ -3002,6 +3002,155 @@ namespace wil
...
+            else 
+            {
+                THROW_IF_FAILED_MSG(hr, "Failed to get next");
+            }
+        }
+    }
+};
+
+
+template<typename IEnumXxx, wistd::enable_if_t<wil::details::has_next_v<IEnumXxx*>, int> = 0>
```
Comment (oldnewthing) [inline include/wil/com.h:3126]: We should allow the caller to specify the custom smart type to `make_range`. 

```cpp
for (auto&& pidl : make_range<IEnumIDList, wil::unique_pidl>(penum.get()))
{
   ...
}
```

Otherwise you have a memory leak because we leak all the pidls.

We should require a smart type if the enumerator produces a pointer.

```cpp
for (auto&& pidl : make_range(penum.get())) // should be illegal - leaks the pidls
{
   ...
}
```



Unfortunately, we cannot distinguish between returning structures with nothing worth destructing
```
struct SafeToJustAbandon
{
   int first, last;
};
```

and structures that require cleanup.

```cpp
struct MustCleanMeUpOrIWillLeak
{
    VARIANT v; // if you fail to VariantClear me, you will leak memory!
};
```

File: include/wil/com.h [hunk 14]
```diff
@@ -3002,6 +3002,155 @@ namespace wil
...
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        // If the result is a COM pointer type (IFoo*), then we use wil::com_ptr<IFoo>. Otherwise, we use the raw pointer type IFoo*.
+        using smart_result = wistd::conditional_t<wistd::is_pointer_v<Result> && wistd::is_base_of_v<::IUnknown, wistd::remove_pointer_t<Result>>,
+            wil::com_ptr<wistd::remove_pointer_t<Result>>, Result>;
+    };
+}
+
+template <typename IEnumType, typename TStoredType = typename details::com_enumerator_traits<IEnumType>::smart_result>
```
Comment (oldnewthing) [inline include/wil/com.h:3053]: We should use [type deduction from the future](https://devblogs.microsoft.com/oldnewthing/20230609-00/?p=108318) to allow the caller to specify only the TStoredType.

```cpp
// cumberwome
auto enumerator = com_iterator<IEnumIDList, wil::unique_idlist>(penum.get());

// better ergonomics
auto enumerator = com_iterator<wil::unique_idlist>(penum.get());
```

File: include/wil/com.h (hunk 13 above)
Comment (asklar) [reply include/wil/com.h:3126]: @oldnewthing Originally this had the TStoredType with a default value of `smart_result`. However, @dunhor mentioned we should sfinae this method, so that requires an optional non-type template param.
So in order to support both I think you'd have to write `make_range<IEnumIDList, int, wil::unique_pidl>` (because putting the TStoredType first causes multiple definition IIRC), see https://github.com/microsoft/wil/pull/398#discussion_r1411281525

Comment (dunhor) [reply include/wil/com.h:3126]: _Technically_ the SFINAE part doesn't need to be a template parameter (though with an `auto` return type, the list of tricks available to you decreases). The original code also had a problem where an additional compilation error was coming in due to a non-dependent (i.e. not a substitution failure) access of the `T::Next` function pointer.

In Raymond's example, `TStoredType` is deduced "as necessary" and doesn't seem to have any of the original issues: https://godbolt.org/z/GTTaa6b7h

Comment (oldnewthing) [reply include/wil/com.h:3126]: Wearing my "samples-driven API design" hat, I figure that we want the calling code to look like this:
```cpp
wil::com_ptr<IEnumUnknown> enumUnknown;
wil::com_ptr<IEnumIDList> enumIDList;

// allowed: deduces IEnumUnknown and com_ptr<IUnknown>
auto enumerator = wil::com_enumerator(enumUnknown.get());

// allowed: deduces IEnumUnknown and uses caller-specified com_ptr_nothrow<IUnknown>
auto enumerator = wil::com_enumerator<wil::com_ptr_nothrow<IUnknown>>(enumUnknown.get());

// allowed: deduces IEnumIDList, uses explicitly-provided unique_idlist
auto enumerator = wil::com_enumerator<wil::unique_idlist>(enumIDLIst.get());

// not allowed - need to specify how to free the IDList
auto enumerator = wil::com_enumerator(enumIDLIst.get());
```
And similarly for `make_range`.

This means that we have
```cpp
template<typename TStoredType = void, typename IEnumXxx>
struct com_enumerator
{
    static_assert(!std::is_same_v<typename com_enumerator_traits::Result, typename com_enumerator_traits::smart_result> || !std::is_same_v<TStoredType, void>, "You must specify a type to receive the enumerated objects.")
};
```
possibly with SFINAE stuff added to fail the com_enumerator at deduction time rather than at instantiation.

Comment (dunhor) [reply include/wil/com.h:3126]: > possibly with SFINAE stuff added to fail the com_enumerator at deduction time rather than at instantiation.

If it's just a struct, there's no real need for SFINAE because there's no overload resolution for structs. It's _occasionally_ useful to SFINAE on specializations, but I can't imagine that being necessary here

File: include/wil/com.h (hunk 14 above)
Comment (asklar) [reply include/wil/com.h:3053]: See #404

File: include/wil/com.h (hunk 13 above)
Comment (asklar) [reply include/wil/com.h:3126]: following up in #404
